# Análisis exploratorio datos cooperativas

## Extraer información de los csv

In [ ]:
import pandas as pd

def load_data(file_path):
    """
    Load data from a CSV file.

    Parameters:
    file_path (str): The path to the CSV file.

    Returns:
    pd.DataFrame: The loaded data as a DataFrame.
    """
    try:
        data = pd.read_csv(file_path)
        return data
    except Exception as e:
        print(f"Error loading data: {e}")
        return None


capital = load_data('../tablas/Capital.csv')
assests = load_data('../tablas/Assets.csv')
managerial = load_data('../tablas/Managerial.csv')
earnings = load_data('../tablas/Earnings.csv')
liquidity = load_data('../tablas/Liquidity.csv')

In [ ]:
capital.head()

,Año,Mes,índice CAMEL,CAJA COOPERATIVA CREDICOOP,CAJA COOPERATIVA PETROLERA,CASA NACIONAL DEL PROFESOR,COOPANTEX COOPERATIVA DE AHORRO Y CREDITO,COOPERATIVA AHORRO Y CREDITO GOMEZ PLATA LTDA.,COOPERATIVA BELEN AHORRO Y CREDITO,COOPERATIVA CALDENSE DEL PROFESOR,...,COOPERATIVA SAN VICENTE DE PAUL LTDA.,COOPERATIVA TELEPOSTAL LTDA,COOPERATIVA TEXAS LTDA,COPERATIVA INDEPENDIENTE DE EMPLEADOS DE ANTIOQUIA,COPERATIVA MULTIACTIVA DE EDUCADORES DE BOYACA,COPROCENVA COOPERATIVA DE AHORRO Y CREDITO,EMPRESA COOPERATIVA DE AHORRO Y CREDITO SIGLO XX LTDA.,FINANCIERA COOPERATIVA COLOMBIANA DE INGENIEROS,GRAN COOPERATIVA DE ENERGIA ELECTRICA Y RECURSOS NATURALES,"MULTIACTIVA EL ROBLE, ENTIDAD COOPERATIVA"
0,2014,1,Quebranto Patrimonial,1.500449,1.310736,1.188087,1.588034,1.481415,2.088982,0.000000,...,1.366898,1.575925,1.430989,1.415437,1.119304,1.367164,1.328639,1.615208,1.634529,2.159444
1,2014,1,Relación Solvencia,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
2,2014,1,Relación entre Aportes sociales mínimos no red...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
3,2014,1,Relación entre el Capital Institucional y el A...,0.145260,0.057027,0.047077,0.126778,0.085107,0.094572,0.000000,...,0.072971,0.157531,0.112465,0.066512,0.031398,0.060938,0.152577,0.116553,0.243769,0.256789
4,2014,2,Quebranto Patrimonial,1.508398,1.310754,1.190007,1.597500,1.480301,2.099399,1.001952,...,1.367030,1.581733,1.440119,1.415271,1.117340,1.370708,1.337403,1.625810,1.637438,2.173214


## Análisis descriptivo incial de los datos

Se requiere analizar los siguientes puntos dentro de cada cooperativa para cada indicador:
* Valores atípicos
* Rango mínimo y máximo
* Distribución general del indicador
* Calcular métricas por años para una cooperativa (media, mediana, desviación estándar, percentiles) para evaluar si mejora o empeora a lo largo de los años.
* Calcular promedio histórico, tendencia (pendiente) y variabilidad de todas las cooperativas en un año específico para detectar cooperativas estables o volátiles.

### Cambiar el formato de las tablas para graficar mejor

In [ ]:
df = pd.concat([capital, assests, managerial, earnings, liquidity], axis=0, ignore_index=True)
df = df[df["Año"] != 2014]
df.head()

,Año,Mes,índice CAMEL,CAJA COOPERATIVA CREDICOOP,CAJA COOPERATIVA PETROLERA,CASA NACIONAL DEL PROFESOR,COOPANTEX COOPERATIVA DE AHORRO Y CREDITO,COOPERATIVA AHORRO Y CREDITO GOMEZ PLATA LTDA.,COOPERATIVA BELEN AHORRO Y CREDITO,COOPERATIVA CALDENSE DEL PROFESOR,...,COOPERATIVA SAN VICENTE DE PAUL LTDA.,COOPERATIVA TELEPOSTAL LTDA,COOPERATIVA TEXAS LTDA,COPERATIVA INDEPENDIENTE DE EMPLEADOS DE ANTIOQUIA,COPERATIVA MULTIACTIVA DE EDUCADORES DE BOYACA,COPROCENVA COOPERATIVA DE AHORRO Y CREDITO,EMPRESA COOPERATIVA DE AHORRO Y CREDITO SIGLO XX LTDA.,FINANCIERA COOPERATIVA COLOMBIANA DE INGENIEROS,GRAN COOPERATIVA DE ENERGIA ELECTRICA Y RECURSOS NATURALES,"MULTIACTIVA EL ROBLE, ENTIDAD COOPERATIVA"
48,2015,1,Quebranto Patrimonial,1.546325,1.230612,1.208915,1.547815,1.447973,2.165285,1.028898,...,1.409897,1.595742,1.456537,1.432761,1.050841,1.379673,1.433775,1.665161,1.617976,2.197900
49,2015,1,Relación Solvencia,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50,2015,1,Relación entre Aportes sociales mínimos no red...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
51,2015,1,Relación entre el Capital Institucional y el A...,0.136617,0.074496,0.051595,0.120289,0.081255,0.079721,0.115896,...,0.085172,0.166284,0.123617,0.061384,0.029987,0.063994,0.158011,0.127597,0.242805,0.250311
52,2015,2,Quebranto Patrimonial,1.551888,1.219469,1.210820,1.547027,1.446249,2.192686,1.030528,...,1.408851,1.607194,1.465172,1.437090,1.045575,1.386981,1.435076,1.668955,1.623512,2.204510


In [ ]:
import matplotlib.pyplot as plt
import ipywidgets as widgets
from IPython.display import display
import pandas as pd

In [ ]:
df_long = pd.read_excel("../tablas/registros_categorizados.xlsx")
df_long = df_long[df_long["ano"] != 2014]
df_long.columns

Index(['ID_registro', 'ID_indicador', 'ID_cooperativa', 'ano', 'mes', 'valor',
       'categoria'],
      dtype='object')

In [ ]:
# lista de indicadores
indicadores = sorted(df_long["ID_indicador"].dropna().unique())

# graficar
def disp_plot_indicador(indicador):
    df_ind = df_long[df_long["ID_indicador"] == indicador]
    
    # dispersión ---
    fig, ax = plt.subplots()
    ax.scatter(df_ind["ano"], df_ind["valor"], alpha=0.5)

    plt.title(f"Distribución de {indicador}")
    plt.suptitle("")
    plt.xlabel("Año")
    plt.ylabel("Valor")
    plt.show()

def hist_plot_indicador(indicador):
    df_ind = df_long[df_long["ID_indicador"] == indicador]
    
    # histograma ---
    plt.hist(df_ind["valor"], bins=10, alpha=0.5)

    plt.title(f"Distribución de {indicador}")
    plt.xlabel("Valor")
    plt.ylabel("Frecuencia")
    plt.show()

def box_plot_indicador(indicador):
    df_ind = df_long[df_long["ID_indicador"] == indicador]
    
    # caja ---
    plt.figure()
    df_ind.boxplot(column="valor", by="ano")

    plt.title(f"Distribución de {indicador} por Año")
    plt.suptitle("")
    plt.xlabel("Año")
    plt.ylabel("Valor")
    plt.show()

# widget dropdown
dropdown = widgets.Dropdown(
    options=indicadores,
    description="Indicador:",
    layout=widgets.Layout(width="60%")
)


In [ ]:
# interacción
widgets.interactive(disp_plot_indicador, indicador=dropdown)

interactive(children=(Dropdown(description='Indicador:', layout=Layout(width='60%'), options=('Activo Producti…

In [ ]:
widgets.interactive(hist_plot_indicador, indicador=dropdown)

interactive(children=(Dropdown(description='Indicador:', index=5, layout=Layout(width='60%'), options=('Activo…

In [ ]:
widgets.interactive(box_plot_indicador, indicador=dropdown)

interactive(children=(Dropdown(description='Indicador:', index=5, layout=Layout(width='60%'), options=('Activo…

In [ ]:
# lista de indicadores
indicadores = sorted(df_long["ID_indicador"].dropna().unique())

# lista de categorias
categorias = sorted(df_long["categoria"].dropna().unique())


# graficar
def disp_plot_indicador(indicador, categoria):

    df_ind = df_long[df_long["ID_indicador"] == indicador]

    if categoria != "Todas":
        df_ind = df_ind[df_ind["categoria"] == categoria]
    
    # dispersión ---
    fig, ax = plt.subplots()
    ax.scatter(df_ind["ano"], df_ind["valor"], alpha=0.5)

    plt.title(f"Distribución de {indicador} - {categoria}")
    plt.suptitle("")
    plt.xlabel("Año")
    plt.ylabel("Valor")
    plt.show()


def hist_plot_indicador(indicador, categoria):

    df_ind = df_long[df_long["ID_indicador"] == indicador]

    if categoria != "Todas":
        df_ind = df_ind[df_ind["categoria"] == categoria]
    
    # histograma ---
    plt.hist(df_ind["valor"], bins=10, alpha=0.5)

    plt.title(f"Distribución de {indicador} - {categoria}")
    plt.xlabel("Valor")
    plt.ylabel("Frecuencia")
    plt.show()


def box_plot_indicador(indicador, categoria):

    df_ind = df_long[df_long["ID_indicador"] == indicador]

    if categoria != "Todas":
        df_ind = df_ind[df_ind["categoria"] == categoria]
    
    # caja ---
    plt.figure()
    df_ind.boxplot(column="valor", by="ano")

    plt.title(f"Distribución de {indicador} por Año - {categoria}")
    plt.suptitle("")
    plt.xlabel("Año")
    plt.ylabel("Valor")
    plt.show()


# -----------------------------
# cálculo de rangos normales
# -----------------------------


# función general para widgets
def analizar(indicador, categoria):

    disp_plot_indicador(indicador, categoria)
    hist_plot_indicador(indicador, categoria)
    box_plot_indicador(indicador, categoria)


# -----------------------------
# widgets
# -----------------------------

dropdown_indicador = widgets.Dropdown(
    options=indicadores,
    description="Indicador:",
    layout=widgets.Layout(width="60%")
)

dropdown_categoria = widgets.Dropdown(
    options=["Todas"] + categorias,
    description="Categoría:",
    layout=widgets.Layout(width="60%")
)


widgets.interactive(
    analizar,
    indicador=dropdown_indicador,
    categoria=dropdown_categoria
)

interactive(children=(Dropdown(description='Indicador:', layout=Layout(width='60%'), options=('Activo Producti…

#### Análisis indicadores:

* Activo productivo: +
    * el rango de valores está aproximadamente entre 0 y 2
    * en el 2014 tiene varios valores atípicos hasta de 60 (revisar eso)
    * dist. no es normal, tiene mucho valores pequeños, asimetría positiva
    * por lo general es un indicador estable
* Estructura de balance: + 
    * comportamiento muy similar a activo productivo pero con un rango mayor, de aproximadamente hasta 10
    * valores atípicos en el 2014 hasta de 240 aprox
* Cobertura de la cartera total en riesgo: +
    * rango de valores entre 0 y 5 aprox
    * asimetría postiva
    * muchos valores atípicos en el 2017
* Cobertura individual de la cartera improductiva para la cartera en riesgo: +
    * varios años con valores atípicos pero más en el 2017
    * rango de valores entre 0 y 10 aprox
    * asimetría positiva

* Margen financiero de operación
    * varios años tienen valores atípicos pero más el 2025
    * rango de valores entre 0.4 y 1.0
    * A diferencia de los anteriores se acerca más a una dist. normal

* Margen operacional: -
    * el 2014 tiene más valores atípicos
    * asimetría negativa
    * rango entre -2 y 0 aprox
* IRL: const
* Calidad por riesgo: +
    * muchos valores atípicos en todos los años, mayormente 2025
    * asimetría positiva
    * rango entre 0 y 0.1 aprox.
* Calidad por riesgo con castigos: +
    * muchos valores atípicos, mayormente 2025
    * asimetría positiva
    * rango entre 0 y 0.05 aprox

* Margen neto:
    * muchos ourliers, mayormente 2025
    * se acerca un poco a la dist. normal
    * rango entre -0.1 a 0.4 aprox

* Relación entre obligaciones financieras y el pasivo total: +
    * muchos outliers, mayormente en el 2025
    * asimetría positiva
    * rango entre 0 y 0.15 aprox

* ROIC:    
    * se acerca un poco a la dist.normal
    * muchos outliers con cantidades parecidas entre los años
    * rango entre -0.2 y 0.4 aprox
* ROE:
    * se acerca un poco a la dist.normal
    * muchos outliers
    * rango entre 0 y 1 aprox
* Quebranto patrimonial:
    * se acerca un poco a la dist.normal
    * muchos outliers, mayormente 2025
    * rango entre 1 y 2 aprox
* Relación de solvencia: const

* Aportes sociales mínimos no reducibles y capital social: -
    * rango entre 0 y 1 aprox
    * asimetría negativa muy leve
    * 2025 con muchos outliers

* Relación entre el Capital Institucional y el Activo Total: +
    * asimetría positivia
    * muchos outliers, mayormente 2025
    * rango entre 0 y 0.2 aprox

### Preguntas

1. ¿Cómo manejamos esos outliers? Porque son muchos.
2. Se puede cambiar la escala de los datos? Como log o exp.

# Notas:

*  Se podría considerar segmentar cooperativas por comportamientos similares (por volatilidad o por un nivel promedio) para identificar perfiles como cooperativas líderes, en crecimiento o inestables
* Se podría complementar el análisis de tendencias de las cooperativas utilizando series de tiempo (y graficarlas)